pkg/controller/seed-controller-manager/addon/addon_controller.go

```diff
@@ -564,22 +592,36 @@ func (r *Reconciler) ensureIsInstalled(ctx context.Context, log *zap.SugaredLogg
 		return fmt.Errorf("failed to create command: %w", err)
 	}
 
-	if addon.Name == "csi" &&
-		cluster.Spec.Cloud.Hetzner != nil &&
-		cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] &&
-		cluster.Annotations[migratedHetznerCSIAddon] != "yes" {
-		// Between v2.22 and v2.23, there was a change to hetzner CSI driver immutable field fsGroupPolicy
-		// as a result, the CSDriver resource has to be redeployed
-		// https://github.com/kubermatic/kubermatic/issues/12429
-		if err := r.migrateHetznerCSIDriver(ctx, log, cluster); err != nil {
-			return fmt.Errorf("failed to migrate CSI Driver: %w", err)
-		}
-		if cluster.Annotations == nil {
-			cluster.Annotations = make(map[string]string)
-		}
-		cluster.Annotations[migratedHetznerCSIAddon] = "yes"
-		if err := r.Update(ctx, cluster); err != nil {
-			log.Errorf("failed to set %q cluster annotation: %w", migratedHetznerCSIAddon, err)
+	if addon.Name == "csi" {
+		if cluster.Spec.Cloud.Hetzner != nil &&
+			cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] &&
+			cluster.Annotations[migratedHetznerCSIAddon] != yes {
```

Comment:
nit: unrelated to your changes but I realized https://github.com/kubermatic/kubermatic/pull/12432 should have come with some expiration date. After everyone upgraded from 2.22 it's essentially just dead code. I feel like we should mark the Hetzner migration code for deprecation and remove perhaps after 2.26 release? Then for vSphere, in 2.27 release?

Reply:
We can remove the Hetzner migration code after 2.24 has been released. I would prefer to leave it in place this close to a release, but yes, we need to clean things up.